internal/server/recon/recognize.go

```diff
@@ -446,24 +455,67 @@ func getNumSpansForContainedIn(spans []*pb.TokenSpans_Span, startIdx int) int {
 	return 0
 }
 
-func combineContainedInSingle(
-	spans []*pb.TokenSpans_Span, startIdx, numSpans int) *pb.TokenSpans_Span {
+func combineTokenSpans(spans []*pb.TokenSpans_Span, startIdx, numSpans int) *pb.TokenSpans_Span {
 	startSpan := spans[startIdx]
-	endSpan := spans[startIdx+numSpans-1]
-
+	
 	res := &pb.TokenSpans_Span{Tokens: startSpan.Tokens}
 	for i := 1; i < numSpans; i++ {
 		res.Tokens = append(res.Tokens, spans[startIdx+i].GetTokens()...)
 	}
 
-	// This map is used to collect all the places for the combined span, with dedup.
+	return combineContainedInTokens(res, spans, startIdx, numSpans)
+}
+
+func getNextPlaceTokenSpan(spans []*pb.TokenSpans_Span, startIdx int) (*pb.TokenSpans_Span, int) {
+	for i := startIdx; i < len(spans); i++ {
+		if len(spans[i].GetPlaces()) > 0 {
+			return spans[i], i
+		}
+	}
+	return nil, -1
+}
+
+func findRecogPlaces(startSpan, nextSpan *pb.TokenSpans_Span) (map[string]*pb.RecogPlace, []string) {
 	dcidToRecogPlaces := map[string]*pb.RecogPlace{}
+	nextContainingPlaces := []string{}
 
 	for _, p1 := range startSpan.GetPlaces() {
 		for _, containingPlace := range p1.GetContainingPlaces() {
-			for _, p2 := range endSpan.GetPlaces() {
+			for _, p2 := range nextSpan.GetPlaces() {
 				if containingPlace == p2.GetDcid() {
 					dcidToRecogPlaces[p1.GetDcid()] = p1
+					nextContainingPlaces = p2.GetContainingPlaces()
+				}
+			}
+		}
+	}
+	return dcidToRecogPlaces, nextContainingPlaces
+}
+
+func combineContainedInTokens(
+	res *pb.TokenSpans_Span, 
+	spans []*pb.TokenSpans_Span, startIdx, numSpans int) *pb.TokenSpans_Span {
+	startSpan := spans[startIdx]
+	nextSpan, nextSpanIndex :=  getNextPlaceTokenSpan(spans, startIdx+1)
+
+	startingDcids := []string{}
+	for _, v := range startSpan.GetPlaces() {
+		startingDcids = append(startingDcids[:], v.GetDcid())
+	}
+
+	// This map is used to collect all the places for the combined span, with dedup.
+	dcidToRecogPlaces, nextContainingPlaces := findRecogPlaces(startSpan, nextSpan)
+	
+	if numSpans == 5 {
+		nextNextSpan, _ := getNextPlaceTokenSpan(spans, nextSpanIndex+1)			
+		for _, possiblePlace := range nextContainingPlaces {
+			for _, actualPlace := range nextNextSpan.GetPlaces() {
+				if actualPlace.GetDcid() == possiblePlace {
+					for _, dcid := range startingDcids {
+						if val, ok := dcidToRecogPlaces[dcid]; ok {
+							val.ContainingPlaces = append(val.ContainingPlaces, possiblePlace)
```

Comment:
I'm not sure if this will happen in practice, but could there be a case where there are 2 different starting places each with a different containing place that matched the next span and so in this case, we wouldn't want to add this last containing place to every starting dcid right?

maybe instead of having nextContainingPlaces be a list, it could be a map of startingPlace to nextContainingPlaces for that starting place.

Reply:
Yeah you're definitely right. Very very unlikely to happen, but also not impossible. I changed it to a map but actually it ended up doing so much iterating through loops. I ended up switching the map to be Containing Place DCID to StartingDCID and I think it's a bit cleaner that way.
So now I ll generate the map as: {country/USA: [geoId/1234, geoId/2345], Earth: [geoId/1234, geoId/2345],...}
So then we can check the nextNextSpan and if it has an entry in the map, we append the key in dcidToRecogPlace for all the starting DCIDs it has.